Bad JSDoc comments leads to generate invalid declaration file

**TypeScript Version:**  3.9.0-dev.20200328

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** `JSDoc` `A required parameter cannot follow an optional parameter.`

**Code**

```ts
/**
 * @param {string} password Passphrase
 * @param {string} [targetRoomId] Room ID to target a specific room.
 * Restores all rooms if omitted.
 * @param {object} opts Optional params such as callbacks
 */
function x(password, targetRoomId, opts) {}
x()
```

**Expected behavior:**
Generate declaration file:
```ts
function x(password: string, targetRoomId?: string, opts?: object): void
```

**Actual behavior:**
Generate declaration file:
```ts
function x(password: string, targetRoomId?: string, opts: object): void
// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^ Notice the question mark `?` has gone
// error TS1016: A required parameter cannot follow an optional parameter.
```

**Playground Link:** https://www.typescriptlang.org/v2/en/play?ts=3.9.0-dev.20200328&useJavaScript=true#code/PQKhFgCgAIWgBADgQwE7ILbQN4GcAuqAlgHYDmAvtCrrgO4D2qAJtAArK2IAW6uAplFgIU6LHkKlK0ANr40ZfvgBKDBhgCSzALrRV66BoAi0fA1MKl0ZNFyJ+AYyIAzIg+io1GAHRC4y-gImQOsAG1CPL1xoF2h1Inx8fmZfGDgkNEwcBgAjACtHfCoGRHxogHlSogYSZAjRTGjcAFcHbmtohzrQnOQHAGtcP2AoZ2aSB3xqkmgADwAKGnomZgAaC1RFFS8tdZKygEocCigoBYOoIA

Hover the mouse on the function `x`, you can see it's signature is inferred as `function x(password: string, targetRoomId?: string, opts: object): void` which is wrong.

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -36841,10 +36841,9 @@ namespace ts {
                         return grammarErrorOnNode(parameter.name, Diagnostics.A_rest_parameter_cannot_have_an_initializer);
                     }
                 }
-                else if (parameter.questionToken) {
+                else if (isOptionalParameter(parameter)) {
                     seenOptionalParameter = true;
-
-                    if (parameter.initializer) {
+                    if (parameter.questionToken && parameter.initializer) {
                         return grammarErrorOnNode(parameter.name, Diagnostics.Parameter_cannot_have_question_mark_and_initializer);
                     }
                 }

--- a/tests/baselines/reference/checkJsdocOptionalParamOrder.errors.txt
+++ b/tests/baselines/reference/checkJsdocOptionalParamOrder.errors.txt
@@ -1,0 +1,14 @@
+tests/cases/conformance/jsdoc/0.js(7,20): error TS1016: A required parameter cannot follow an optional parameter.
+
+
+==== tests/cases/conformance/jsdoc/0.js (1 errors) ====
+    // @ts-check
+    /**
+     * @param {number} a
+     * @param {number} [b]
+     * @param {number} c
+     */
+    function foo(a, b, c) {}
+                       ~
+!!! error TS1016: A required parameter cannot follow an optional parameter.
+    

--- a/tests/baselines/reference/checkJsdocOptionalParamOrder.js
+++ b/tests/baselines/reference/checkJsdocOptionalParamOrder.js
@@ -1,0 +1,18 @@
+//// [0.js]
+// @ts-check
+/**
+ * @param {number} a
+ * @param {number} [b]
+ * @param {number} c
+ */
+function foo(a, b, c) {}
+
+
+//// [0.js]
+// @ts-check
+/**
+ * @param {number} a
+ * @param {number} [b]
+ * @param {number} c
+ */
+function foo(a, b, c) { }

--- a/tests/baselines/reference/checkJsdocOptionalParamOrder.symbols
+++ b/tests/baselines/reference/checkJsdocOptionalParamOrder.symbols
@@ -1,0 +1,13 @@
+=== tests/cases/conformance/jsdoc/0.js ===
+// @ts-check
+/**
+ * @param {number} a
+ * @param {number} [b]
+ * @param {number} c
+ */
+function foo(a, b, c) {}
+>foo : Symbol(foo, Decl(0.js, 0, 0))
+>a : Symbol(a, Decl(0.js, 6, 13))
+>b : Symbol(b, Decl(0.js, 6, 15))
+>c : Symbol(c, Decl(0.js, 6, 18))
+

--- a/tests/baselines/reference/checkJsdocOptionalParamOrder.types
+++ b/tests/baselines/reference/checkJsdocOptionalParamOrder.types
@@ -1,0 +1,13 @@
+=== tests/cases/conformance/jsdoc/0.js ===
+// @ts-check
+/**
+ * @param {number} a
+ * @param {number} [b]
+ * @param {number} c
+ */
+function foo(a, b, c) {}
+>foo : (a: number, b?: number, c: number) => void
+>a : number
+>b : number
+>c : number
+

--- a/tests/baselines/reference/jsdocParseBackquotedParamName.errors.txt
+++ b/tests/baselines/reference/jsdocParseBackquotedParamName.errors.txt
@@ -1,0 +1,20 @@
+tests/cases/conformance/jsdoc/a.js(5,18): error TS1016: A required parameter cannot follow an optional parameter.
+
+
+==== tests/cases/conformance/jsdoc/a.js (1 errors) ====
+    /**
+     * @param {string=} `args`
+     * @param `bwarg` {?number?}
+     */
+    function f(args, bwarg) {
+                     ~~~~~
+!!! error TS1016: A required parameter cannot follow an optional parameter.
+    }
+    
+==== tests/cases/conformance/jsdoc/ts.ts (0 errors) ====
+    /**
+     * @param `arg` - this is fine
+     */
+    function g(arg: string) {
+    }
+    

--- a/tests/cases/conformance/jsdoc/checkJsdocOptionalParamOrder.ts
+++ b/tests/cases/conformance/jsdoc/checkJsdocOptionalParamOrder.ts
@@ -1,0 +1,11 @@
+﻿// @allowJS: true
+// @suppressOutputPathCheck: true
+
+// @filename: 0.js
+// @ts-check
+/**
+ * @param {number} a
+ * @param {number} [b]
+ * @param {number} c
+ */
+function foo(a, b, c) {}

